docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+```
+
+## Desired directory structure
+
+We want to define a standard layout and structure for Kubernetes API types.
+Below is a first draft, with explanation afterwards. We may try to move the
+relevant parts of the legacy `pkg/api/v1` api into the new structure, although
+this api doesn't have a group.
+
+```
+apis
+└── group
+    ├── install
+    ├── types.go
+    ├── v1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
+    │   │   │   └── manual_conversion_overrides.go
+    │   │   ├── defaulting
+    │   │   │   ├── generated_defaulters.go
+    │   │   │   └── generated_default_stubs.go
+    │   │   └── validation
+    │   │       ├── generated_validation.go
+    │   │       └── generated_validation_stubs.go
+    │   └── types.go
+    ├── v1beta1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
+    │   │   │   └── manual_conversion_overrides.go
+    │   │   ├── defaulting
+    │   │   │   ├── generated_defaulters.go
+    │   │   │   └── generated_default_stubs.go
+    │   │   └── validation
+    │   │       ├── generated_validation.go
+    │   │       └── generated_validation_stubs.go
+    │   └── types.go
+    └── v2alpha1
+        ├── install
+        │   ├── conversion
+        │   │   ├── generated_conversion.go
+        │   │   └── manual_conversion_overrides.go
+        │   ├── defaulting
+        │   │   ├── generated_defaulters.go
+        │   │   └── generated_default_stubs.go
+        │   └── validation
+        │       ├── generated_validation.go
+        │       └── generated_validation_stubs.go
+        └── types.go
+```
+
+### Package: `apis/`
+
+Multiple groups may live in the same tree. Each group gets its own directory
+under `apis/`. We'll assume an `apis/` directory of this structure in all tools.
+
+TBD: Either test code will live here, or no code.
+
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a types.go file contains an "internal" version of the group's
+objects. This file is used as a destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
+little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers the v1 version of this group, its
+conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/install/conversion`
```

Comment:
I don't like that this is nested under install - it just doesn't seem obvious.  I also am not sure about the idea that this will register if imported.  Do we have cases where we want to register conversions but NOT defaulting (for example)?

What I was thinking was something like:

`apis/group/v1/install/install.go` imports `apis/group/v1/{conversion,defaults,validation}` and calls `{conversion,defaults,validation}.Register(args)`.  This leaves a minimum of boilerplate gunk inside `{conversion,defaults,validation}/` and accumulates it all in `install/`.  Then we can de-boilerplate that if needed.

If we're going to spend energy making  small well-defined pkgs, let's really try to push the noisy code out of them.